### ISSUE TITLE
Revisit shared process and main process services

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#114418]
Follow up from our work on shared process using message ports: we currently have a `IMainProcessService` and `ISharedProcessService` that expose methods to `getChannel` and `registerChannel`. I think the names are misleading and should probably include the IPC part, or we find an alternative.

The original necessity for adding these services was to get them properly injected and not create services eagerly on startup.

Up for ideas how to improve this.

**Comment by bpasero** [comment]
My proposal would be to simply include `IPC` in the names of these services, but I think @joaomoreno you might have had more in mind to hide implementation specific details?

**Comment by joaomoreno** [comment]
After discussion with @bpasero, here's what we should do:

- Remove all injections of `ISharedProcessService` and `IMainProcessService` from misc classes: channels shouldn't be used across the workbench, except in implementing `IChannel` and `IChannelClient` interfaces
- There should be `registerSharedProcessService` et al methods which register services coming in from the shared process via channels. Optionally, `Channel` and `ChannelClient` ctors can be passed as options, in case the ProxyChannel util can't be used.

```ts
registerSharedProcessService(ILocalizationsService, 'localizations', Channel?, ChannelClient?);
```

**Comment by Tyriar** [comment]
I'm a little worried about the merge headaches this sort of thing will cause when trying to get https://github.com/microsoft/vscode/pull/116373 in. Something a little unique about that is that there are 2 levels of proxy (window to sharedprocess to ptyhost) which I don't think we're doing anywhere yet.

**Comment by Tyriar** [comment]
The `ProxyChannel` utils are great, but this double proxy concept seemed to fall apart when handing events going back to the window:

https://github.com/microsoft/vscode/blob/17018593c7ca8065f9422a2fb19cd9677aabe089/src/vs/platform/terminal/electron-browser/localPtyService.ts#L18-L19

**Comment by joaomoreno** [comment]
Using the `ProxyChannel` is by no means necessary. You can always resort to implementing your own `Channel`/`ChannelClient` pair, as it's done in some places, eg https://github.com/microsoft/vscode/blob/2cd7a70d629fd5cac500341174291b0bfd120a6c/src/vs/platform/url/common/urlIpc.ts#L12-L37

**Comment by joaomoreno** [comment]
@bpasero: After reviewing our approach with @sandy081, we figured out that in [some of his code](https://github.com/microsoft/vscode/blob/1e8edfea6ed7660f582b752e1587abcf264c32ed/src/vs/workbench/services/userDataSync/electron-sandbox/userDataSyncStoreManagementService.ts#L27-L28) it will be very hard not to depend on `ISharedProcessService`. Namely, extending from an abstract class which depends on service injecting in the ctor is the restricting factor.

So we came up with something different. The real problem isn't that `ISharedProcessService` is accessibly from anywhere, but rather that it exposes low-level channels via `getChannel`. Here's the current interface:

```ts
export interface ISharedProcessService {
	getChannel(channelName: string): IChannel;
	registerChannel(channelName: string, channel: IServerChannel<string>): void;
}
```

What if we would move to an interface which would force people to use services in their application layer code? Something like:

```ts
export interface RemoteServiceCtor<T> {
  new(channel: IChannel): T;
}

export interface ISharedProcessService {
	getService<T>(channelName: string, ctor: RemoteServiceCtor): T;
	registerChannel(channelName: string, channel: IServerChannel<string>): void;
}
```

That would avoid that people save `IChannel` instances and use them directly.

In any case, we should still have the `registerSharedProcessService` (et al) util for the common case.

**Comment by joaomoreno** [comment]
I've pushed some code which should highlight the two patterns, custom IPC and _automagic_ proxy IPC:

**Proxy:** [diagnosticsService.ts](https://github.com/microsoft/vscode/blob/141a47120336066c28f552d86e554f823c9895f1/src/vs/workbench/services/diagnostics/electron-sandbox/diagnosticsService.ts)
**Custom:** [updateService.ts](https://github.com/microsoft/vscode/blob/141a47120336066c28f552d86e554f823c9895f1/src/vs/workbench/services/update/electron-sandbox/updateService.ts)

What we will further do is reduce those files to a simpler pattern.

**Proxy:**

```ts
registerSharedProcessService(IDiagnosticsService, DiagnosticsService);
```

**Custom:**

```ts
registerMainProcessService(IUpdateService, 'update', { channelClientCtor: UpdateChannelClient });
```

---

That way, the direct use of `IMainProcessService` and `ISharedProcessService` should be heavily reduced, as well as the necessary `// @ts-ignore` hack.

**Comment by bpasero** [comment]
@Tyriar 

> I'm a little worried about the merge headaches this sort of thing will cause when trying to get #116373 in. Something a little unique about that is that there are 2 levels of proxy (window to sharedprocess to ptyhost) which I don't think we're doing anywhere yet.

I am not sure I fully understand this, aren't we in the end shooting for a solution that allows direct window to ptyhost communication via `MessagePort`? In that case I would expect some kind of `ITerminalPtyHostService` that the renderer can depend on to talk to the host directly.

@joaomoreno 

> That would avoid that people save `IChannel` instances and use them directly. @sandy081 could still reach out to use `ISharedProcessService`, but he would be forced to write a `ChannelClient` class for his remote service, keeping all the unsafe typing contained in that class.

Any particular reason why the proxy cannot be used in this case? The proxy supports events, so even [this code](https://github.com/microsoft/vscode/blob/1e8edfea6ed7660f582b752e1587abcf264c32ed/src/vs/workbench/services/userDataSync/electron-sandbox/userDataSyncStoreManagementService.ts#L27-L28) can use the proxy. 

In general I like the idea of hiding channels, but shouldn't then `registerChannel` also go away?

> That way, the direct use of IMainProcessService and ISharedProcessService should be heavily reduced, as well as the necessary // @ts-ignore workaround.

Yeah I like that. Btw I noticed you moved `ISharedProcessService` and `IMainProcessService` to `common`. My original intent was to keep them in `electron-sandbox` as clear signal that you cannot depend on these from the browser. I would strongly suggest to keep these interfaces in `electron-sandbox` so that we are not accidentally dragging those interfaces into the web where they do not make a lot of sense.

**Comment by joaomoreno** [comment]
> Any particular reason why the proxy cannot be used in this case? The proxy supports events, so even [this code](https://github.com/microsoft/vscode/blob/1e8edfea6ed7660f582b752e1587abcf264c32ed/src/vs/workbench/services/userDataSync/electron-sandbox/userDataSyncStoreManagementService.ts#L27-L28) can use the proxy.

Yup, it's related to him extending from a base class which requires other services. I can talk it thought in a screen sharing session.

> I would strongly suggest to keep these interfaces in `electron-sandbox` so that we are not accidentally dragging those interfaces into the web where they do not make a lot of sense.

Yeah I had a weird feeling pulling them out. Your argument made me go back to my tracks. Reverted that.

> In general I like the idea of hiding channels, but shouldn't then `registerChannel` also go away?

Yes we need to come up with a similar alternative for that too.

---

Here's the first commit showing an ideal solution: https://github.com/microsoft/vscode/commit/b2e68d70b8dd30b565e6b181fac6ec47b7e00f3d

Notice the usage on both situations:

- Proxy: [diagnosticsService.ts](https://github.com/microsoft/vscode/blob/b2e68d70b8dd30b565e6b181fac6ec47b7e00f3d/src/vs/workbench/services/diagnostics/electron-sandbox/diagnosticsService.ts)
- Custom channel client: [updateService.ts](https://github.com/microsoft/vscode/blob/b2e68d70b8dd30b565e6b181fac6ec47b7e00f3d/src/vs/workbench/services/update/electron-sandbox/updateService.ts)

Neither ever sees instances of `IMainProcessService` or `ISharedProcessService`.

**Comment by bpasero** [comment]
> Yeah I had a weird feeling pulling them out. Your argument made me go back to my tracks. Reverted that.

Thanks

> Here's the first commit showing an ideal solution: b2e68d7

Looks great, should go ahead and adopt this for all our services 👍 

One nit: it looks like `updateService.ts` lives in `workbench` but should probably move to `platform`. There is a `browser` variant that is very workbench specific, but I would still think that it is better to have them all together in one place?

**Comment by joaomoreno** [comment]
> One nit: it looks like `updateService.ts` lives in `workbench` but should probably move to `platform`. There is a `browser` variant that is very workbench specific, but I would still think that it is better to have them all together in one place?

Yeah no preference here. The distinction between platform and workbench services isn't really super clear in my mind. Maybe you can clarify?

> Looks great, should go ahead and adopt this for all our services 👍

Awesome. Now that we have a clear direction, I'll move this to March debt week. 👌

**Comment by bpasero** [comment]
Sounds good, let me summarise the work here also given that I will merge https://github.com/microsoft/vscode/issues/116318 into this issue:
* adopt `registerMainProcessRemoteService` and `registerSharedProcessRemoteService` everywhere
* try to use the proxy as much as possible but fallback to existing IPC overrides if custom solution is needed (related https://github.com/microsoft/vscode/issues/116191)
* consider to move services closer together into one `platform` folder instead of spreading across `platform` and `workbench`

**Comment by bpasero** [comment]
An inventory where we still implement the channel manually:
* [ ] `src/vs/platform/debug/common/extensionHostDebugIpc.ts`
* [ ] `src/vs/platform/download/common/downloadIpc.ts`
* [ ] `src/vs/platform/driver/node/driver.ts`
* [ ] `src/vs/platform/extensionManagement/common/extensionManagementIpc.ts`
* [ ] `src/vs/platform/extensionRecommendations/electron-sandbox/extensionRecommendationsIpc.ts`
* [ ] `src/vs/platform/log/common/logIpc.ts`
* [ ] `src/vs/platform/request/common/requestIpc.ts`
* [ ] `src/vs/platform/storage/electron-main/storageIpc.ts`
* [ ] `src/vs/platform/telemetry/common/telemetryIpc.ts`
* [ ] `src/vs/platform/update/common/updateIpc.ts`
* [ ] `src/vs/platform/url/common/urlIpc.ts`
* [ ] `src/vs/platform/userDataSync/common/userDataSyncIpc.ts`
* [ ] `src/vs/platform/userDataSync/common/userDataSyncServiceIpc.ts`
* [ ] `src/vs/workbench/services/search/node/searchApp.ts`

Should sign off here when we adopted `ProxyChannel`

Btw unfortunately I had to move `ISharedProcessService` into `workbench` because of dependencies to the workbench lifecycle service. That means the IPC services are not nicely grouped in `platform` anymore but I also think that is fair because the shared process service is meant to only be used from renderer.

**Comment by bpasero** [comment]
@joaomoreno while adopting `registerMainProcessRemoteService` and `registerSharedProcessRemoteService` for some of mine, I noticed that I have no way of configuring the proxy. For this case I need it though:

https://github.com/microsoft/vscode/blob/ea048b877175ae2736ee5ff93273490cf57c4e68/src/vs/workbench/services/workspaces/electron-sandbox/workspacesService.ts#L21-L21

Should we allow to pass `channelClientCtor | ProxyOptions` so that it is more obvious how to configure the proxy when no channel client is passed in?

**Comment by joaomoreno** [comment]
> Should we allow to pass `channelClientCtor | ProxyOptions` so that it is more obvious how to configure the proxy when no channel client is passed in?

Yeah sounds good.

**Comment by bpasero** [comment]
Ok I can do that change in my branch.

**Comment by bpasero** [comment]
Oh nevermind, turns out I cannot use that because the properties I need are coming from other services. I will have to keep workspaces service and native host service as they are.

**Comment by joaomoreno** [comment]
Closing it for now.
